### PR TITLE
[Workflow] Choose which Workflow events should be dispatched

### PR DESCRIPTION
**Pull request by stewartmalik**
See issue #14062

### CODE REVIEW

File: workflow.rst [hunk 1]
```diff
@@ -452,6 +444,146 @@ missing a title::
 
     The optional second argument of ``setBlocked()`` was introduced in Symfony 5.1.
 
+.. _workflow-usage-choose-events:
+
+Choosing which Events to dispatch
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (noniagriconomie) [inline workflow.rst:451]: i would add a `.. versionadded:: 5.2` directive here :) 

```
.. versionadded:: 5.2
Ability to choose which events to dispatch was introduced in Symfony 5.2.
```

File: workflow.rst [hunk 2]
```diff
@@ -452,6 +444,146 @@ missing a title::
...
+You are able to specify which events will be fired when performing each
```
Comment (noniagriconomie) [inline workflow.rst:452]:
```suggestion
You are able to specify which events (does not apply to Guard event) will be fired when performing each
```

File: workflow.rst [hunk 3]
```diff
@@ -452,6 +444,146 @@ missing a title::
...
+transition by passing an array of workflow events to the ``events_to_dispatch``
```
Comment (noniagriconomie) [inline workflow.rst:453]: before showing config code, i would describe briefly the 3 options available:
- null => all events
- [] => no events
- [...|...] => specific events

**Comment by stewartmalik** [comment]
@noniagriconomie thanks for the review. Have updated the PR with the requested changes.

File: workflow.rst [hunk 4]
```diff
@@ -452,6 +444,155 @@ missing a title::
...
+
+You are able to specify which events (does not apply to Guard event) will be
+fired when performing each transition by passing an array of workflow events
+to the ``events_to_dispatch`` configuration option.
+
+Valid options for ``events_to_dispatch`` are:
+
+    * ``null`` - all events are dispatched
+    * ``[]`` - no events are dispatched (except the Guard event)
+    * ``['workflow.leave', 'workflow.completed']`` - specific events are dispatched
```
Comment (OskarStark) [inline workflow.rst:464]:
```suggestion
    * ``['workflow.leave', 'workflow.completed']`` - only specific events are dispatched
```

File: workflow.rst [hunk 5]
```diff
@@ -452,6 +444,155 @@ missing a title::
...
+                        'workflow.leave',
+                        'workflow.completed',
+                    ],
+                    // ...
+                ],
+            ],
+        ]);
+
+To specify that no events will be dispatched pass an empty array to the
+configuration option. Guard Events are still dispatched in this instance.
```
Comment (OskarStark) [inline workflow.rst:516]:
```suggestion
configuration option.

.. note::

    Guard Events are still dispatched in this instance.
```

What do you think of making this sentence more prominent?

File: workflow.rst [hunk 6]
```diff
@@ -452,6 +444,155 @@ missing a title::
...
+    } catch (LogicException $exception) {
+        // ...
+    }
+
+Choosing to disable an event for a specific transition will take precedence
+over any events specified in the workflow configuration. In the above example
+the ``workflow.leave`` event will not be fired, even if it has been specified
+as an event to be dispatched for all transitions in the Workflow configuration.
+
+.. versionadded:: 5.2
```
Comment (OskarStark) [inline workflow.rst:588]:
```suggestion
.. versionadded:: 5.1
```

File: workflow.rst [hunk 7]
```diff
@@ -452,6 +444,155 @@ missing a title::
...
+            Workflow::DISABLE_LEAVE_EVENT => true,
+        ]);
+    } catch (LogicException $exception) {
+        // ...
+    }
+
+Choosing to disable an event for a specific transition will take precedence
+over any events specified in the workflow configuration. In the above example
+the ``workflow.leave`` event will not be fired, even if it has been specified
+as an event to be dispatched for all transitions in the Workflow configuration.
```
Comment (OskarStark) [inline workflow.rst:586]: this looks wrong according to the code above, does it needs to be `false` in the code? 🤔

Comment (OskarStark) [inline workflow.rst:586]:
```suggestion
as an event to be dispatched for all transitions in the workflow configuration.
```

**Review by OskarStark — CHANGES_REQUESTED** [review]
Some minor comments

File: workflow.rst [hunk 8]
```diff
@@ -452,6 +444,155 @@ missing a title::
 
     The optional second argument of ``setBlocked()`` was introduced in Symfony 5.1.
 
+.. _workflow-usage-choose-events:
+
+Choosing which Events to dispatch
```
Comment (OskarStark) [inline workflow.rst:449]:
```suggestion
Choosing which Events to Dispatch
```

File: workflow.rst (hunk 6 above)
Comment (noniagriconomie) [reply workflow.rst:588]: unsure of this, feature was merged into master and sf5.1 is already released, so feature is for 5.2 no?
or did i misunderstood something?
@OskarStark

File: workflow.rst (hunk 5 above)
Comment (noniagriconomie) [reply workflow.rst:516]: also i would put this note on top
after the `Valid options for ``events_to_dispatch`` are:`
as it is documented ` (except the Guard event)`

Comment (stewartmalik) [reply workflow.rst]: I've added the note further up as @noniagriconomie suggested, does that look alright?

File: workflow.rst [hunk 9]
```diff
@@ -452,6 +444,159 @@ missing a title::
 
     The optional second argument of ``setBlocked()`` was introduced in Symfony 5.1.
 
+.. _workflow-usage-choose-events:
```
Comment (OskarStark) [inline workflow.rst:447]: it is unused, isn't it?

File: workflow.rst [hunk 10]
```diff
@@ -452,6 +444,159 @@ missing a title::
...
+the ``workflow.leave`` event will not be fired, even if it has been specified
+as an event to be dispatched for all transitions in the workflow configuration.
+
+.. versionadded:: 5.1
+
+    The ``Workflow::DISABLE_ANNOUNCE_EVENT`` constant was introduced in Symfony 5.1.
+
+.. versionadded:: 5.2
+
+    The constants for other events, such as ``Workflow::DISABLE_LEAVE_EVENT`` were introduced in Symfony 5.2.
```
Comment (OskarStark) [inline workflow.rst:598]: Could we list alle constants instead of using "such as ...". This would make the docs more discoverable and searchable for the users.

File: workflow.rst (hunk 7 above)
Comment (stewartmalik) [reply workflow.rst]: I believe it should be passed as `true` (or technically can be passed as any value that's not `null`) however perhaps @lyrixx can confirm?

File: workflow.rst [hunk 11]
```diff
@@ -452,6 +444,163 @@ missing a title::
...
+
+You are also able to explicitly disable a specific event from being fired
+when applying a transition::
+
+    use App\Entity\BlogPost;
+    use Symfony\Component\Workflow\Exception\LogicException;
+
+    $post = new BlogPost();
+
+    $workflow = $this->container->get('workflow.blog_publishing');
```
Comment (noniagriconomie) [inline workflow.rst:574]: i think you should put this code inside a controller or so, because here the $this->container seems unrelated

**Review by OskarStark — APPROVED** [review]

File: workflow.rst (hunk 7 above)
Comment (lyrixx) [reply workflow.rst]:
```
Workflow::DISABLE_LEAVE_EVENT => true
```
Is the way to disable the listener

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Tests/WorkflowTest.php#L433-L435

**Review by lyrixx — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Stewart thanks a lot for this contribution! It was great, specially for a first-time contribution! Congrats on that :tada:

Note: while merging we did some tweaks and rewords to merge some contents to make the section a bit more concise ... but most of your original contribution was left intact. Cheers!
